### ISSUE TITLE
Gridview rendering/event binding issues when updating by ajax (patch included)

### ISSUE THREAD

**Issue by qiangxue** [yiisoft/yii#270]
<b>What steps will reproduce the problem?</b>
1. create gridview with filter &amp; ajax enabled
2. change filter value and trigger ajax update by pressing enter
3. change another filter value before the update has completed 
4. add your own custom filter and bind (or live) some javascript  

What do you see ?
1. gridview update script replaces the complete grid after ajax complete, clearing any values changed after the request was made and removing the event handlers from the filter.
2. only the grid is replaced, new event bindings from the ajax request are not being executed.

<b>What version of the product are you using? On what operating system?</b>
Yii 1.1.9

Please see attached patch for jquery.yiigridview.js
- preserves current filter (including event handlers) before replacing the grid, allowing for example to use inputwidgets like autocomplete in the filter form.
- stores current update request in settings and cancels it before starting  a new update request.

Thank you for your ongoing support &amp; development of this great PHP framework!<p>Migrated from http://code.google.com/p/yii/issues/detail?id=3117</p><hr/><h2>earlier comments</h2>

<p><strong>mdomba said, at 2012-01-18T07:48:38.000Z:</strong></p>

<p>Can you create a small test app to test this?

Your solution presumes that you are again loading the jquery.yiigridview.js on every ajax call. The current idea is to prevent that so that all delegated events remains there.</p><p><strong>w.westenbrink said, at 2012-01-18T09:59:31.000Z:</strong></p>

<p>Thank you for your quick reply!

As you know gridview currently retrieves the entire page when updating but only replaces elements with ids from settings.ajaxUpdate. 

I updated the patch to make it more clear i only preserve the filter when available before replacing the grid completely:

if ($data.find('.filters').length > 0 && $('.filters', updateId).length > 0) {
    $data.find('.filters').replaceWith($('.filters', updateId).detach());
}

Please note that while event delegation works perfectly without this change, added data from the autocomplete plugin for example is lost when the filter is replaced.</p><p><strong>w.westenbrink said, at 2012-01-18T10:10:12.000Z:</strong></p>

<p>Another idea would be to replace settings.ajaxUpdate with more complex array of id's ( excluding filters) or to add a new setting called ajaxPreserve.
 </p><p><strong>mdomba said, at 2012-01-18T11:18:52.000Z:</strong></p>

<p>gridview retrieves what the developer sends him... the default app generated by Gii returns the entire page but that is only for simplicity... in real use you should return just the data needed to be refreshed.

I still don't understand why you want to preserve the filter... wouldn't that be returned by the ajax call too? Custom events binded should remain there like the gridview one...

I would like to see an example application with the problem you want to solve... can you create a test case?</p><p><strong>w.westenbrink said, at 2012-01-18T13:11:06.000Z:</strong></p>

<p>thank you for your insight in this.

I only use renderPartial on my view containing the CGridView widget when updating via ajax. 

A more optimal case which also seems to solves my problem is changing the "The data that needs to be refreshed" (settings.ajaxUpdate) to:
- tbody element of the grid's table
- .pager element
- .summary element 
- .keys element

correct ?

I want to preserve the filter input elements because jquery ui widgets like autocomplete store jquery data() in these input elements.

Leaves "stores current update request in settings and cancels it before starting  a new update request", do you want me to create a new issue for that ?</p><p><strong>mdomba said, at 2012-01-18T13:28:59.000Z:</strong></p>

<p>before going into further discussion and/or opening new tickets... 

as I explained already I would like to see an example application with this problem, that we can work on, to solve the problem... can you create a test app for this ?</p><p><strong>w.westenbrink said, at 2012-01-18T13:39:49.000Z:</strong></p>

<p>thank you for your support, it seems my filters problem is solved by specifying more precise ajaxUpdate setting.

I would like to setup a test app for "stores current update request in settings and cancels it before starting  a new update request", what do you need ?</p><p><strong>mdomba said, at 2012-01-18T13:47:23.000Z:</strong></p>

<p>a working test case "controller - model - view" that shows the problem you want to fix... 

What I want is that you show me the problem... not the solution... as there can be "maybe" another way to solve it</p><p><strong>w.westenbrink said, at 2012-01-18T14:01:28.000Z:</strong></p>

<p>I understand, please allow me some time to setup the test case. do you need access to the source code to ?</p><p><strong>mdomba said, at 2012-01-18T14:08:17.000Z:</strong></p>

<p>of course, there is no need to hurry :D

for the test case... you don't need to set a live (online) test case... just post here the files (or send them to my email)... but if it's easier for you to set a live example than it's fine for me, too...</p>
